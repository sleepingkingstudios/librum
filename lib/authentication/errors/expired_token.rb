# frozen_string_literal: true

module Authentication::Errors
  # Abstract error returned for an expired authentication token.
  class ExpiredToken < Authentication::Errors::InvalidToken
    # Short string used to identify the type of error.
    TYPE = 'authentication.errors.expired_token'

    private

    def default_message
      'expired authentication token'
    end
  end
end
