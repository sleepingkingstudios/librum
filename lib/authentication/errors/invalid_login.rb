# frozen_string_literal: true

require 'librum/core/errors/authentication_error'

module Authentication::Errors
  # Error returned for a failed username+password login.
  class InvalidLogin < Librum::Core::Errors::AuthenticationError
    # Short string used to identify the type of error.
    TYPE = 'authentication.errors.invalid_login'

    # @param errors [Stannum::Errors] the errors object, if any.
    def initialize(errors: nil)
      super(
        errors:  errors,
        message: 'invalid username or password'
      )

      @errors = errors
    end

    # @return [Stannum::Errors] the errors object, if any.
    attr_reader :errors

    private

    def as_json_data
      return {} unless errors

      { 'errors' => errors.as_json }
    end
  end
end
