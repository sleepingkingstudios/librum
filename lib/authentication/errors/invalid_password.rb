# frozen_string_literal: true

require 'cuprum'

module Authentication::Errors
  # Error returned when a password does not match the encrypted value.
  class InvalidPassword < Cuprum::Error
    # Short string used to identify the type of error.
    TYPE = 'authentication.errors.invalid_password'

    # @param message [String] message describing the nature of the error.
    def initialize(message: nil, **options)
      super(
        message: message || default_message,
        **options
      )
    end

    private

    def default_message
      'password does not match encrypted value'
    end
  end
end
