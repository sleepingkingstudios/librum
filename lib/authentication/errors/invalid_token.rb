# frozen_string_literal: true

require 'cuprum'

module Authentication::Errors
  # Abstract error returned for an invalid authentication token.
  class InvalidToken < Cuprum::Error
    # Short string used to identify the type of error.
    TYPE = 'authentication.errors.invalid_token'

    # @param message [String] Message describing the nature of the error.
    def initialize(message: nil, **options)
      super(
        message: message || default_message,
        **options
      )
    end

    private

    def default_message
      'invalid authentication token'
    end
  end
end
