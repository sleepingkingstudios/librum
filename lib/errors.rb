# frozen_string_literal: true

# Namespace for application-level errors.
module Errors; end
