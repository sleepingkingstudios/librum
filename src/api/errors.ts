export const expiredSessionError = 'authentication.errors.expiredToken';

export const invalidParametersError = 'cuprum.rails.errors.invalidParameters';

export const invalidPasswordError = 'authentication.errors.invalidPassword';

export const missingTokenError = 'authentication.errors.missingToken';

export const notFoundError = 'cuprum.collections.errors.notFound';
