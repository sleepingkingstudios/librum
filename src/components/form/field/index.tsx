import * as React from 'react';
import {
  capitalize,
  last,
  map,
  snakeCase,
} from 'lodash';
import {
  useField,
  useFormikContext,
} from 'formik';
import type {
  FieldHookConfig,
  FieldInputProps,
} from 'formik';

import type { UpToTwelveColumns } from '@components/types';
import { joinClassNames } from '@utils/react-utils';
import { FormInput } from '../input';
import type { FormStatus } from '../types';
import { getServerErrors } from '../utils';

interface IFormFieldProps {
  className?: string;
  cols?: UpToTwelveColumns;
  id?: string;
  label?: string | false;
  name: string;
  type?: string;
}

interface IFormLabelProps {
  id: string;
  label: string | false;
}

interface IRenderComponentProps {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  field: FieldInputProps<any>;
  id?: string;
  type: string;
}

const columnSpanClassName = ({ cols }: { cols: number }): string | null => {
  if (cols === null) { return null; }

  return `col-span-${cols}`;
};

const convertNameToId = (name: string): string => (
  name.replace(/\]/g, '').split('[').map(snakeCase).join('_')
);

const convertNameToLabel = (name: string): string => {
  const segment: string = last(name.split(/\[/));
  const trimmed: string = snakeCase(segment.replace(/\]/, ''));

  return map(trimmed.split('_'), capitalize).join(' ');
};

const renderComponent = ({
  field,
  id = null,
  type,
}: IRenderComponentProps): JSX.Element => (
  <FormInput field={field} id={id} type={type} />
);

const renderErrors = ({ errors }: { errors: string[] }): JSX.Element => {
  if (errors.length === 0) { return; }

  return (<FormFieldErrors errors={errors} />);
};

const renderLabel = ({ id, label }: IFormLabelProps): JSX.Element => {
  if (label === false) {
    return null;
  }

  return (
    <label className="form-label" htmlFor={id}>{ label }</label>
  );
};

const FormFieldErrors = ({ errors }: { errors: string[] }): JSX.Element => (
  <span className="form-field-errors">{ errors.join(', ') }</span>
);

export const FormField = ({
  className = null,
  cols = null,
  id = null,
  label = null,
  name,
  type = 'text',
}: IFormFieldProps): JSX.Element => {
  const configuredId = id || convertNameToId(name);
  const configuredLabel =
    label === false ? label : (label || convertNameToLabel(name));
  const joinedClassName = joinClassNames(
    'form-field',
    columnSpanClassName({ cols }),
    className,
  );
  const fieldConfig: FieldHookConfig<string> = {
    name,
  };
  const [field] = useField(fieldConfig);
  const context = useFormikContext();
  const status = context.status as FormStatus;
  const errors = getServerErrors({ name, status });

  return (
    <div className={joinedClassName}>
      { renderLabel({ id: configuredId, label: configuredLabel }) }

      { renderComponent({ id: configuredId, field, type }) }

      { renderErrors({ errors }) }
    </div>
  );
};
