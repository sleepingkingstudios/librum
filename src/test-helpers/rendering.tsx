import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
