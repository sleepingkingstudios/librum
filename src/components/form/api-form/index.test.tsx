import * as React from 'react';
import { faRadiation } from '@fortawesome/free-solid-svg-icons';

import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { ApiForm } from './index';
import {
  responseWithError,
  responseWithStatus,
} from '@api';
import type {
  Refetch,
  Response,
} from '@api';
import { invalidParametersError } from '@api/errors';
import type { Animations } from '@components/types';
import { render } from '@test-helpers/rendering';
import { FormField } from '../field';
import type {
  FormErrors,
  InvalidParametersError,
} from '../types';

describe('<ApiForm />', () => {
  const response: Response = responseWithStatus({ status: 'success' });
  const refetch: jest.MockedFunction<Refetch> = jest.fn(
    () => new Promise(resolve => resolve(response))
  );

  beforeEach(() => {
    refetch.mockClear();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <ApiForm initialValues={{}} refetch={refetch} response={response}>
        <button type="submit">Submit</button>
      </ApiForm>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when the response is loading', () => {
    const response: Response = responseWithStatus({ status: 'loading' });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ApiForm initialValues={{}} refetch={refetch} response={response}>
          <button type="submit">Submit</button>
        </ApiForm>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with loading configuration', () => {
      const options = {
        loadingAnimation: 'pulse' as Animations,
        loadingIcon: faRadiation,
        loadingMessage: 'Activating Reactor...',
      };

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <ApiForm
            initialValues={{}}
            refetch={refetch}
            response={response}
            {...options}
          >
            <button type="submit">Submit</button>
          </ApiForm>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('when the user submits the form', () => {
    it('should perform the request', async () => {
      const { getByRole } = render(
        <ApiForm initialValues={{}} refetch={refetch} response={response}>
          <button type="submit">Submit</button>
        </ApiForm>
      );
      const button = getByRole('button');

      await userEvent.click(button);

      expect(refetch).toHaveBeenCalledWith({ body: {} });
    });
  });

  describe('with className: value', () => {
    const className = 'custom-form';

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ApiForm
          className={className}
          initialValues={{}}
          refetch={refetch}
          response={response}
        >
          <button type="submit">Submit</button>
        </ApiForm>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with form fields', () => {
    const Fields = (): JSX.Element => (
      <React.Fragment>
        <label>
          Launch Site
          <FormField name="launchSite" />
        </label>

        <label>
          Mission Name
          <FormField name="mission.name" />
        </label>
      </React.Fragment>
    );
    const defaultValues = {
      launchSite: '',
      mission: {
        name: '',
      }
    };

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ApiForm
          initialValues={defaultValues}
          refetch={refetch}
          response={response}
        >
          <Fields />

          <button type="submit">Submit</button>
        </ApiForm>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the user submits the form', () => {
      it('should perform the request', async () => {
        const { getByRole } = render(
          <ApiForm
            initialValues={defaultValues}
            refetch={refetch}
            response={response}
          >
            <Fields />

            <button type="submit">Submit</button>
          </ApiForm>
        );
        const button = getByRole('button');

        await userEvent.click(button);

        expect(refetch).toHaveBeenCalledWith({ body: defaultValues });
      });

      describe('when the request returns a response with API errors', () => {
        const errors: FormErrors = {
          'launchSite': [{
            data: {},
            message: "can't be blank",
            path: ['launchSite'],
            type: 'test.errorType',
          }],
        };
        const error: InvalidParametersError = {
          data: { errors },
          message: 'invalid request parameters',
          type: invalidParametersError,
        };
        const failureResponse = responseWithError({
          error,
          errorType: error.type,
        });

        beforeEach(() => {
          refetch.mockImplementationOnce(
            () => new Promise(resolve => resolve(failureResponse)),
          );
        });

        it('should match the snapshot', async () => {
          const { asFragment, getByRole, rerender } = render(
            <ApiForm
              initialValues={defaultValues}
              refetch={refetch}
              response={response}
            >
              <Fields />

              <button type="submit">Submit</button>
            </ApiForm>
          );
          const button = getByRole('button');

          await userEvent.click(button);

          rerender(
            <ApiForm
              initialValues={defaultValues}
              refetch={refetch}
              response={failureResponse}
            >
              <Fields />

              <button type="submit">Submit</button>
            </ApiForm>
          );

          expect(asFragment()).toMatchSnapshot();
        });
      });
    });

    describe('when the user enters form data', () => {
      describe('when the user submits the form', () => {
        const expectedValues = {
          launchSite: 'KSC',
          mission: {
            name: 'Warlock IV',
          },
        };

        it('should perform the request', async () => {
          const { getByLabelText, getByRole } = render(
            <ApiForm
              initialValues={defaultValues}
              refetch={refetch}
              response={response}
            >
              <Fields />

              <button type="submit">Submit</button>
            </ApiForm>
          );
          const button = getByRole('button');

          await userEvent.type(getByLabelText('Launch Site'), 'KSC');

          await userEvent.type(getByLabelText('Mission Name'), 'Warlock IV');

          await userEvent.click(button);

          expect(refetch).toHaveBeenCalledWith({ body: expectedValues });
        });
      });
    });
  });
});
