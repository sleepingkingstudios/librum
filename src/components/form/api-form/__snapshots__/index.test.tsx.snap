// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ApiForm /> should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<ApiForm /> when the response is loading should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <div
      class="form-loading-overlay absolute grid h-full place-content-center w-full"
    >
      <div
        class="inline-block text-center"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-gear animate-spin motion-reduce:animate-none h-10 w-10"
          data-icon="gear"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M481.9 166.6c3.2 8.7 .5 18.4-6.4 24.6l-30.9 28.1c-7.7 7.1-11.4 17.5-10.9 27.9c.1 2.9 .2 5.8 .2 8.8s-.1 5.9-.2 8.8c-.5 10.5 3.1 20.9 10.9 27.9l30.9 28.1c6.9 6.2 9.6 15.9 6.4 24.6c-4.4 11.9-9.7 23.3-15.8 34.3l-4.7 8.1c-6.6 11-14 21.4-22.1 31.2c-5.9 7.2-15.7 9.6-24.5 6.8l-39.7-12.6c-10-3.2-20.8-1.1-29.7 4.6c-4.9 3.1-9.9 6.1-15.1 8.7c-9.3 4.8-16.5 13.2-18.8 23.4l-8.9 40.7c-2 9.1-9 16.3-18.2 17.8c-13.8 2.3-28 3.5-42.5 3.5s-28.7-1.2-42.5-3.5c-9.2-1.5-16.2-8.7-18.2-17.8l-8.9-40.7c-2.2-10.2-9.5-18.6-18.8-23.4c-5.2-2.7-10.2-5.6-15.1-8.7c-8.8-5.7-19.7-7.8-29.7-4.6L69.1 425.9c-8.8 2.8-18.6 .3-24.5-6.8c-8.1-9.8-15.5-20.2-22.1-31.2l-4.7-8.1c-6.1-11-11.4-22.4-15.8-34.3c-3.2-8.7-.5-18.4 6.4-24.6l30.9-28.1c7.7-7.1 11.4-17.5 10.9-27.9c-.1-2.9-.2-5.8-.2-8.8s.1-5.9 .2-8.8c.5-10.5-3.1-20.9-10.9-27.9L8.4 191.2c-6.9-6.2-9.6-15.9-6.4-24.6c4.4-11.9 9.7-23.3 15.8-34.3l4.7-8.1c6.6-11 14-21.4 22.1-31.2c5.9-7.2 15.7-9.6 24.5-6.8l39.7 12.6c10 3.2 20.8 1.1 29.7-4.6c4.9-3.1 9.9-6.1 15.1-8.7c9.3-4.8 16.5-13.2 18.8-23.4l8.9-40.7c2-9.1 9-16.3 18.2-17.8C213.3 1.2 227.5 0 242 0s28.7 1.2 42.5 3.5c9.2 1.5 16.2 8.7 18.2 17.8l8.9 40.7c2.2 10.2 9.4 18.6 18.8 23.4c5.2 2.7 10.2 5.6 15.1 8.7c8.8 5.7 19.7 7.7 29.7 4.6l39.7-12.6c8.8-2.8 18.6-.3 24.5 6.8c8.1 9.8 15.5 20.2 22.1 31.2l4.7 8.1c6.1 11 11.4 22.4 15.8 34.3zM242 336a80 80 0 1 0 0-160 80 80 0 1 0 0 160z"
            fill="currentColor"
          />
        </svg>
        <br />
        Loading...
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<ApiForm /> when the response is loading with loading configuration should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <div
      class="form-loading-overlay absolute grid h-full place-content-center w-full"
    >
      <div
        class="inline-block text-center"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-radiation animate-pulse motion-reduce:animate-none h-10 w-10"
          data-icon="radiation"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M215 186.7c-23.9 13.8-40 39.7-40 69.3L31 256c-17.7 0-32.2-14.4-30-31.9C9.7 154 46.8 92.7 100.3 52c14.1-10.7 33.8-5.3 42.7 10l72 124.7zM255 336c14.6 0 28.2-3.9 40-10.7l72 124.8c8.8 15.3 3.7 35.1-12.6 41.9c-30.6 12.9-64.2 20-99.4 20s-68.9-7.1-99.4-20c-16.3-6.9-21.4-26.6-12.6-41.9l72-124.8c11.8 6.8 25.4 10.7 40 10.7zm224-80l-144 0c0-29.6-16.1-55.5-40-69.3L367 62c8.8-15.3 28.6-20.7 42.7-10c53.6 40.7 90.6 102 99.4 172.1c2.2 17.5-12.4 31.9-30 31.9zM257 208a48 48 0 1 1 -2 96 48 48 0 1 1 2-96z"
            fill="currentColor"
          />
        </svg>
        <br />
        Activating Reactor...
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<ApiForm /> with className: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="custom-form relative"
  >
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<ApiForm /> with form fields should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <label>
      Launch Site
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="launch_site"
        >
          Launch Site
        </label>
        <input
          class="form-input"
          id="launch_site"
          name="launchSite"
          type="text"
          value=""
        />
      </div>
    </label>
    <label>
      Mission Name
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="mission_name"
        >
          Mission Name
        </label>
        <input
          class="form-input"
          id="mission_name"
          name="mission.name"
          type="text"
          value=""
        />
      </div>
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<ApiForm /> with form fields when the user submits the form when the request returns a response with API errors should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <label>
      Launch Site
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="launch_site"
        >
          Launch Site
        </label>
        <input
          class="form-input"
          id="launch_site"
          name="launchSite"
          type="text"
          value=""
        />
        <span
          class="form-field-errors"
        >
          can't be blank
        </span>
      </div>
    </label>
    <label>
      Mission Name
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="mission_name"
        >
          Mission Name
        </label>
        <input
          class="form-input"
          id="mission_name"
          name="mission.name"
          type="text"
          value=""
        />
      </div>
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
